app/src/main/java/fr/free/nrw/commons/settings/SettingsFragment.java

```diff
@@ -157,35 +172,99 @@ private void prepareLanguages() {
             languageNameWithCodeList.add(languageNamesList.get(i) + "[" + languageCodesList.get(i) + "]");
         }
 
-        CharSequence[] languageNames = languageNameWithCodeList.toArray(new CharSequence[0]);
-        CharSequence[] languageCodes = languageCodesList.toArray(new CharSequence[0]);
+        final CharSequence[] languageNames = languageNamesList.toArray(new CharSequence[0]);
+        final CharSequence[] languageCodes = languageCodesList.toArray(new CharSequence[0]);
         // Add all languages and languages codes to lists preference as pair
-        langListPreference.setEntries(languageNames);
-        langListPreference.setEntryValues(languageCodes);
-
-        // Gets current language code from shared preferences
-        String languageCode = getCurrentLanguageCode();
-        if (languageCode.equals("")) {
-            // If current language code is empty, means none selected by user yet so use phone local
-            langListPreference.setValue(Locale.getDefault().getLanguage());
-        } else {
-            // If any language is selected by user previously, use it
-            langListPreference.setValue(languageCode);
+
+        if(keyListPreference.equals("appUiDefaultLanguagePref")){
+            appUiLanguageListPreference.setEntries(languageNames);
+            appUiLanguageListPreference.setEntryValues(languageCodes);
+
+            // Gets current language code from shared preferences
+            final String languageCode = getCurrentLanguageCode(keyListPreference);
+            assert languageCode != null;
+            if (languageCode.equals("")){
+                // If current language code is empty, means none selected by user yet so use phone local
+                appUiLanguageListPreference.setValue(Locale.getDefault().getLanguage());
+            } else {
+                // If any language is selected by user previously, use it
+                appUiLanguageListPreference.setValue(languageCode);
+            }
+
+            appUiLanguageListPreference.setOnPreferenceChangeListener((preference, newValue) -> {
+                final String userSelectedValue = (String) newValue;
+                setLocale(Objects.requireNonNull(getActivity()),userSelectedValue);
+                saveLanguageValue(userSelectedValue, keyListPreference);
+                getActivity().recreate();
+                final Intent intent = new Intent(getActivity(), MainActivity.class);
+                startActivity(intent);
+                return true;
+            });
+        }else if(keyListPreference.equals("descriptionDefaultLanguagePref")){
+            descriptionLanguageListPreference.setEntries(languageNames);
+            descriptionLanguageListPreference.setEntryValues(languageCodes);
+
+            // Gets current language code from shared preferences
+            final String languageCode = getCurrentLanguageCode(keyListPreference);
```

Comment:
This line is the same as line 184, right?
How about factorizing by putting it before the `if`, if you think that makes sense?

Reply:
Yes, I can do that. I missed that.